ftdc/custom_format.go

```diff
@@ -0,0 +1,549 @@
+package ftdc
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+
+	"go.viam.com/rdk/logging"
+)
+
+// epsilon is a small value for determining whether a float is 0.0.
+const epsilon = 1e-9
+
+type Schema struct {
+	// A `Datum`s data is a map[string]any. Even if two datum's have maps with the same keys, we do
+	// not assume ranging over the map will yield the same order. Thus we explicitly write down an
+	// order to walk the map of new data values.
+	//
+	// mapOrder is only used for writing out new data points.
+	mapOrder []string
+
+	// fieldOrder is flattened list of strings representing individual metrics. Fields use a
+	// dot-notation to represent structure/nesting. E.g: "leftMotor.PowerPct".
+	fieldOrder []string
+}
+
+// writeSchema writes down names for metrics in the form of a json array. All subsequent calls to
```

Comment:
You could put some of this documentation in a `doc.go` file for the package. Might be nice to have a more obvious place for the description of the format.